### ISSUE TITLE
workspace.onDidChangeConfiguration fired when no configuration changed

### ISSUE THREAD

**Issue by jackieli-tes** [neoclide/coc.nvim#3675]
## versions

vim version: NVIM v0.6.1
node version: v17.5.0
coc.nvim version: 0.0.80-c7d7756089
coc.nvim directory: /home/jackieli/.dotfiles/nvim/plugged/coc.nvim
term: tmux
platform: linux

## Log of coc.nvim

2022-03-02T18:52:04.618 INFO (pid:1633017) [configurations] - Add folder configuration from cwd: /home/jackieli/gomod/df-workflow/.vim/coc-settings.json
2022-03-02T18:52:04.619 INFO (pid:1633017) [configurations] - Change folder configuration from cwd to: /home/jackieli/gomod/df-workflow/.vim/coc-settings.json
2022-03-02T18:52:04.779 INFO (pid:1633017) [configurations] - Add folder configuration from file: /home/jackieli/.dotfiles/.vim/coc-settings.json
2022-03-02T18:52:04.782 INFO (pid:1633017) [configurations] - Change folder configuration to: /home/jackieli/.dotfiles/.vim/coc-settings.json
2022-03-02T18:52:04.831 INFO (pid:1633017) [services] - registered service "languageserver.ccls"
2022-03-02T18:52:04.831 INFO (pid:1633017) [services] - registered service "languageserver.dockerfile"
2022-03-02T18:52:04.831 INFO (pid:1633017) [services] - registered service "languageserver.bash"
2022-03-02T18:52:04.913 INFO (pid:1633017) [coc-git] - Looking for git in: git
2022-03-02T18:52:04.930 INFO (pid:1633017) [services] - registered service "highlight"
2022-03-02T18:52:04.968 INFO (pid:1633017) [services] - registered service "json"
2022-03-02T18:52:04.968 INFO (pid:1633017) [services] - Json language server state change: stopped => starting
2022-03-02T18:52:05.639 INFO (pid:1633017) [plugin] - coc.nvim initialized with node: v17.5.0 after 934ms
2022-03-02T18:52:05.639 INFO (pid:1633017) [services] - registered service "tsserver"
2022-03-02T18:52:05.745 INFO (pid:1633017) [services] - Json language server state change: starting => running
2022-03-02T18:52:05.748 INFO (pid:1633017) [services] - service json started
2022-03-02T18:52:07.724 WARN (pid:1633017) [workspace] - workspace.createStatusBarItem is deprecated, please use window.createStatusBarItem instead. 
    at Object.t.initStatusBar (/home/jackieli/.dotfiles/coc/extensions/node_modules/coc-spell-checker/out/index.js:1:155873)
    at Object.t.activate (/home/jackieli/.dotfiles/coc/extensions/node_modules/coc-spell-checker/out/index.js:1:23099)
2022-03-02T18:52:07.736 INFO (pid:1633017) [attach] - receive notification: highlight []
2022-03-02T18:52:07.737 INFO (pid:1633017) [attach] - receive notification: highlight []
2022-03-02T18:52:07.738 INFO (pid:1633017) [attach] - receive notification: highlight []
2022-03-02T18:52:07.738 INFO (pid:1633017) [attach] - receive notification: highlight []
2022-03-02T18:52:08.899 INFO (pid:1633017) [attach] - receive notification: doAutocmd [ 2 ]
2022-03-02T18:52:09.013 INFO (pid:1633017) [configurations] - Change folder configuration to: /home/jackieli/gomod/df-workflow/.vim/coc-settings.json
2022-03-02T18:52:09.067 INFO (pid:1633017) [services] - registered service "go"
2022-03-02T18:52:09.067 INFO (pid:1633017) [services] - gopls state change: stopped => starting
2022-03-02T18:52:09.080 INFO (pid:1633017) [core-watchman] - watchman watching project: /home/jackieli/gomod/df-workflow
2022-03-02T18:52:09.155 INFO (pid:1633017) [services] - gopls state change: starting => running
2022-03-02T18:52:09.158 INFO (pid:1633017) [services] - service go started
2022-03-02T18:52:10.272 INFO (pid:1633017) [attach] - receive notification: highlight []
2022-03-02T18:52:10.583 INFO (pid:1633017) [configurations] - Change folder configuration to: /home/jackieli/.dotfiles/.vim/coc-settings.json
2022-03-02T18:52:10.600 INFO (pid:1633017) [coc-go] - affected config go.goplsOptions
2022-03-02T18:52:10.601 INFO (pid:1633017) [attach] - receive notification: doAutocmd [ 2 ]
2022-03-02T18:52:11.405 INFO (pid:1633017) [attach] - receive notification: highlight []
2022-03-02T18:52:13.178 INFO (pid:1633017) [attach] - receive notification: showInfo []

## Describe the bug

I'm debugging an extension and it has the following code:

```
const restartConfigs = [
  "go.goplsArgs",
  "go.goplsOptions",
  "go.goplsPath",
  "go.goplsUseDaemon",
]

   let client = new LanguageClient(...)
   ...
    workspace.onDidChangeConfiguration(async (e) => {
      if (restartConfigs.find((k) => e.affectsConfiguration(k))) {
        logger.info('affected config', restartConfigs.find((k) => e.affectsConfiguration(k)))
        // await client.stop()
        // client.restart()
      }
    }),
```

So the logic here is if I see a config change that require restarting the lsp client, I restart it.

However, I'm seeing when there is no config change but the event is still fired. Take the relevant part from above log:

```
2022-03-02T18:52:09.013 INFO (pid:1633017) [configurations] - Change folder configuration to: /home/jackieli/gomod/df-workflow/.vim/coc-settings.json
2022-03-02T18:52:09.067 INFO (pid:1633017) [services] - registered service "go"
2022-03-02T18:52:10.583 INFO (pid:1633017) [configurations] - Change folder configuration to: /home/jackieli/.dotfiles/.vim/coc-settings.json
2022-03-02T18:52:10.600 INFO (pid:1633017) [coc-go] - affected config go.goplsOptions
```

In the above scenario, I have opened my actual project in `$HOME/gomod/df-workflow` and then I want to check my `:CocConfig` which is located in the `$HOME/.dotfiles`. As soon as I do that, without modify anything, the config change event is fired.

In addition, I made sure my `$HOME/.dotfiles` are not a "workspace folder" by adding:

```
  "workspace.ignoredFolders": ["$HOME", "$HOME/.dotfiles"],
```

So in conclusion: I suppose this isn't really a bug, but rather a question: 

**How do we detect the config change for the current project?**

Should I receive a config change for my workspace / project when I open a buffer located outside of the current workspace?

/cc @josa42

**Comment by chemzqm** [comment]
`"workspace.ignoredFolders"` only exclude the folder from add to workspace folders, doesn't affect configurations.
But it should not cause configuration if that's your user configuration file https://github.com/neoclide/coc.nvim/blob/ultisnips/src/configuration/index.ts#L128

**Comment by chemzqm** [comment]
> Should I receive a config change for my workspace / project when I open a buffer located outside of the current workspace

Maybe not, but then we have to check the configuration when current buffer changes for all the handlers, seems a good idea.
